File: previous-gen-apps/focus-kraliki/backend/app/schemas/workflow.py
from pydantic import BaseModel
from typing import List, Optional, Dict, Any
from datetime import datetime


class WorkflowStep(BaseModel):
    step: int
    action: str
    estimatedMinutes: int
    dependencies: List[int] = []
    type: str = "manual"  # manual or automated


class WorkflowTemplateCreate(BaseModel):
    name: str
    description: Optional[str] = None
    category: Optional[str] = None
    icon: Optional[str] = None
    steps: List[Dict[str, Any]]
    tags: List[str] = []
    isPublic: bool = False
    isSystem: bool = False


class WorkflowTemplateUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    category: Optional[str] = None
    icon: Optional[str] = None
    steps: Optional[List[Dict[str, Any]]] = None
    tags: Optional[List[str]] = None
    isPublic: Optional[bool] = None


class WorkflowTemplateResponse(BaseModel):
    id: str
    userId: Optional[str]
    name: str
    description: Optional[str]
    category: Optional[str]
    icon: Optional[str]
    steps: List[Dict[str, Any]]
    totalEstimatedMinutes: Optional[int]
    tags: List[str]
    isPublic: bool
    isSystem: bool
    usageCount: int
    createdAt: datetime

    class Config:
        from_attributes = True


class WorkflowListResponse(BaseModel):
    templates: List[WorkflowTemplateResponse]
    total: int


class WorkflowExecuteRequest(BaseModel):
    templateId: str
    customTitle: Optional[str] = None
    priority: Optional[int] = None
    startDate: Optional[str] = None
    additionalTags: List[str] = []


class WorkflowExecuteResponse(BaseModel):
    success: bool
    message: str
    parentTaskId: str
    createdTasks: List[Dict[str, Any]]
    totalTasks: int

File: previous-gen-apps/focus-kraliki/frontend/src/lib/stores/workflow.ts
/**
 * Workflow Store
 * Svelte writable store for workflow template management
 */

import { writable } from 'svelte/store';
import { api } from '$lib/api/client';
import { logger } from '$lib/utils/logger';

export interface WorkflowStep {
    step: number;
    action: string;
    estimatedMinutes: number;
    dependencies: number[];
    type: 'manual' | 'automated';
}

export interface WorkflowTemplate {
    id: string;
    userId?: string;
    name: string;
    description: string;
    category: string;
    icon?: string;
    steps: WorkflowStep[];
    totalEstimatedMinutes: number;
    tags: string[];
    isPublic: boolean;
    isSystem: boolean;
    usageCount: number;
    createdAt: string;
}

export interface WorkflowState {
    templates: WorkflowTemplate[];
    categories: string[];
    isLoading: boolean;
    error: string | null;
}

const initialState: WorkflowState = {
    templates: [],
    categories: [],
    isLoading: false,
    error: null
};

function createWorkflowStore() {
    const { subscribe, set, update } = writable<WorkflowState>(initialState);

    return {
        subscribe,

        async loadTemplates(params?: { category?: string }) {
            update((state) => ({ ...state, isLoading: true, error: null }));
            try {
                const response: any = await api.workflow.listTemplates(params);
                const templates = response.templates || [];

                update((state) => ({
                    ...state,
                    templates,
                    isLoading: false
                }));

                return { success: true, templates };
            } catch (error: any) {
                const errorMessage = error.detail || 'Failed to load templates';
                update((state) => ({
                    ...state,
                    isLoading: false,
                    error: errorMessage
                }));
                return { success: false, error: errorMessage };
            }
        },

        async loadCategories() {
            try {
                const response: any = await api.workflow.getCategories();
                const categories = response.categories || [];

                update((state) => ({
                    ...state,
                    categories
                }));

                return { success: true, categories };
            } catch (error: any) {
                logger.error('Failed to load categories', error);
                return { success: false, error: error.detail };
            }
        },

        async createTemplate(templateData: any) {
            update((state) => ({ ...state, isLoading: true, error: null }));
            try {
                const newTemplate: any = await api.workflow.createTemplate(templateData);

                update((state) => ({
                    ...state,
                    templates: [newTemplate, ...state.templates],
                    isLoading: false
                }));

                return { success: true, template: newTemplate };
            } catch (error: any) {
                const errorMessage = error.detail || 'Failed to create template';
                update((state) => ({
                    ...state,
                    isLoading: false,
                    error: errorMessage
                }));
                return { success: false, error: errorMessage };
            }
        },

        async generateTemplate(description: string, category?: string) {
            update((state) => ({ ...state, isLoading: true, error: null }));
            try {
                const newTemplate: any = await api.workflow.generate({ description, category });

                update((state) => ({
                    ...state,
                    templates: [newTemplate, ...state.templates],
                    isLoading: false
                }));

                return { success: true, template: newTemplate };
            } catch (error: any) {
                const errorMessage = error.detail || 'Failed to generate template';
                update((state) => ({
                    ...state,
                    isLoading: false,
                    error: errorMessage
                }));
                return { success: false, error: errorMessage };
            }
        },

        async executeWorkflow(templateId: string, options: { startDate?: string; priority?: number; customTitle?: string }) {
            try {
                const response = await api.workflow.execute({
                    templateId,
                    startDate: options.startDate,
                    priority: options.priority,
                    customTitle: options.customTitle
                }) as any;
                return { success: true, ...response };
            } catch (error: any) {
                return { success: false, error: error.detail || 'Failed to execute workflow' };
            }
        },

        async deleteTemplate(templateId: string) {
            try {
                await api.workflow.deleteTemplate(templateId);
                update((state) => ({
                    ...state,
                    templates: state.templates.filter(t => t.id !== templateId)
                }));
                return { success: true };
            } catch (error: any) {
                return { success: false, error: error.detail || 'Failed to delete template' };
            }
        }
    };
}

export const workflowStore = createWorkflowStore();
